arcgis-ios-sdk-samples/Display information/Sketch on the map/SketchViewController.swift

```diff
@@ -16,81 +16,108 @@ import UIKit
-        case 2:// freehand polyline
-            self.sketchEditor.start(with: nil, creationMode: .freehandPolyline)
-            
-        case 3:// polygon
-            self.sketchEditor.start(with: nil, creationMode: .polygon)
-            
-        case 4:// freehand polygon
-            self.sketchEditor.start(with: nil, creationMode: .freehandPolygon)
-            
-        default:
-            break
+    @IBAction func addGeometryButtonTapped(_ sender: UIBarButtonItem) {
+        // Create an alert controller for the action sheets.
+        let alertController = UIAlertController(title: "Select a creation mode", message: nil, preferredStyle: .actionSheet)
+        // Create an action for each creation mode and add it to the alert controller.
+        creationModes.forEach { name, mode in
+            let action = UIAlertAction(title: name, style: .default) { [weak self] _ in
+                self?.statusLabel.text = "\(name) selected."
+                self?.sketchEditor.start(with: nil, creationMode: mode)
+            }
+            alertController.addAction(action)
         }
+        // Add "cancel" item.
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
+        alertController.addAction(cancelAction)
         
-        self.mapView.sketchEditor = self.sketchEditor
+        // Present the action sheets when the add button is tapped.
+        alertController.popoverPresentationController?.barButtonItem = addBarButtonItem
+        present(alertController, animated: true)
     }
     
     @IBAction func undo() {
-        if self.sketchEditor.undoManager.canUndo { // extra check, just to be sure
-            self.sketchEditor.undoManager.undo()
-        }
+        // Check if there are actions to undo.
+        guard sketchEditor.undoManager.canUndo else { return }
+        sketchEditor.undoManager.undo()
     }
     
     @IBAction func redo() {
-        if self.sketchEditor.undoManager.canRedo { // extra check, just to be sure
-            self.sketchEditor.undoManager.redo()
-        }
+        // Check if there are actions to redo.
+        guard sketchEditor.undoManager.canRedo else { return }
+        sketchEditor.undoManager.redo()
     }
     
     @IBAction func clear() {
         self.sketchEditor.clearGeometry()
     }
+    
+    // MARK: - Views
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        // Add an observer to udate UI when needed.
+        barItemObserver = NotificationCenter.default.addObserver(forName: .AGSSketchEditorGeometryDidChange, object: sketchEditor, queue: nil, using: { [unowned self] _ in
+            // Enable/disable UI elements appropriately.
+            self.undoBarButtonItem.isEnabled = self.sketchEditor.undoManager.canUndo
+            self.redoBarButtonItem.isEnabled = self.sketchEditor.undoManager.canRedo
+            self.clearBarButtonItem.isEnabled = self.sketchEditor.geometry.map { !$0.isEmpty } ?? false
```

Comment:
```suggestion
            undoBarButtonItem.isEnabled = sketchEditor.undoManager.canUndo
            redoBarButtonItem.isEnabled = sketchEditor.undoManager.canRedo
            clearBarButtonItem.isEnabled = sketchEditor.geometry.map { !$0.isEmpty } ?? false
```
Since `self` is unowned.